File: Usul/App/Application.cpp

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007, Arizona State University
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//  Created by: Adam Kubach
//
///////////////////////////////////////////////////////////////////////////////

#include "Usul/App/Application.h"
#include "Usul/CommandLine/Arguments.h"
#include "Usul/User/Directory.h"
#include "Usul/Predicates/FileExists.h"

#include <algorithm>

using namespace Usul::App;


///////////////////////////////////////////////////////////////////////////////
//
//  Initialize static data members.
//
///////////////////////////////////////////////////////////////////////////////

Application * Application::_instance ( 0x0 );


///////////////////////////////////////////////////////////////////////////////
//
//  Create instance.
//
///////////////////////////////////////////////////////////////////////////////

Application& Application::instance()
{
  if ( 0x0 == _instance )
  {
    _instance = new Application;
  }

  return *_instance;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Constructor.
//
///////////////////////////////////////////////////////////////////////////////

Application::Application() :
	_splashImage(),
  _vendor(),
  _program()
{
}


///////////////////////////////////////////////////////////////////////////////
//
//  Destructor.
//
///////////////////////////////////////////////////////////////////////////////

Application::~Application()
{
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get the directory where the icons live.
//
///////////////////////////////////////////////////////////////////////////////

const std::string Application::iconDirectory() const
{
	const std::string dir ( Usul::CommandLine::Arguments::instance().directory() );
  
#ifdef __APPLE__
  std::string path ( dir + "/../Resources/icons/" );
#else
	std::string path ( dir + "/icons/" );
#endif

  return path;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Set the splash image name.
//
///////////////////////////////////////////////////////////////////////////////

void Application::splashImage ( const std::string& image )
{
	_splashImage = image;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get the splash image name.
//
///////////////////////////////////////////////////////////////////////////////

const std::string & Application::splashImage() const
{
	return _splashImage;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get the path to the splash screen image.
//
///////////////////////////////////////////////////////////////////////////////

std::string Application::splashImagePath() const
{
	return this->iconDirectory() + this->splashImage();
}


///////////////////////////////////////////////////////////////////////////////
//
//  Set the vender.
//
///////////////////////////////////////////////////////////////////////////////

void Application::vendor ( const std::string& vendor )
{
  _vendor = vendor;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get the vender.
//
///////////////////////////////////////////////////////////////////////////////

const std::string& Application::vendor() const
{
  return _vendor;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Set the program name.
//
///////////////////////////////////////////////////////////////////////////////

void Application::program ( const std::string& program )
{
  _program = program;
}


///////////////////////////////////////////////////////////////////////////////
//
//  Get the program name.
//
///////////////////////////////////////////////////////////////////////////////

const std::string& Application::program() const
{
  return _program;
}

File: Legacy/DotNet/Helios/Lib/Application.cs

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2006, Perry L Miller IV
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//
///////////////////////////////////////////////////////////////////////////////

namespace CadKit.Helios
{
  public class Application
  {
    /// <summary>
    /// Constructor
    /// </summary>
    private Application()
    {
    }

    /// <summary>
    /// Single instance.
    /// </summary>
    public static Application Instance
    {
      get
      {
        lock ("CadKit.Helios.Application.Instance")
        {
          if (null == _instance)
            _instance = new Application();
          return _instance;
        }
      }
    }

    /// <summary>
    /// Set/get the name of the program. Gets set from the assembly (program) name.
    /// </summary>
    public string Name
    {
      get { lock (_mutex) { return _name; } }
      set { lock (_mutex) { _name = value; } }
    }

    /// <summary>
    /// Set/get the main form.
    /// </summary>
    public CadKit.Helios.MainForm MainForm
    {
      get { lock (_mutex) { return _mainForm; } }
      set { lock (_mutex) { _mainForm = value; } }
    }

    /// <summary>
    /// Get the directory where this assembly resides.
    /// </summary>
    public string Directory
    {
      get
      {
        lock (_mutex)
        {
          string path = System.Windows.Forms.Application.ExecutablePath;
          string dir = System.IO.Path.GetDirectoryName(path);
          return dir;
        }
      }
    }

    /// <summary>
    /// Get the assembly file name.
    /// </summary>
    public string FileName
    {
      get
      {
        lock (_mutex)
        {
          return System.Reflection.Assembly.GetEntryAssembly().GetName().Name + ".exe";
        }
      }
    }

    /// <summary>
    /// Get the icons directory.
    /// </summary>
    public string IconDir
    {
      get
      {
        lock (_mutex)
        {
          string dir = this.Directory + "/icons";
          return dir;
        }
      }
    }


    /// <summary>
    /// Get/Set the splash screen image.
    /// </summary>
    public string SplashImage
    {
      get { lock (_mutex) { return _splashImage; } }
      set { lock (_mutex) { _splashImage = value; } }
    }

    /// <summary>
    /// Data members.
    /// </summary>
    private static Application _instance = null;
    private string _name = System.Reflection.Assembly.GetEntryAssembly().GetName().Name; // Handles ".vshost";
    private string _splashImage = null;
    private CadKit.Helios.MainForm _mainForm = null;
    private object _mutex = new object();
  }
}
